Change layout of add widget screen.

rogerl50 opened the pull request:
This changes the add widget screen to a grid and makes the crowded panes much larger.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: https://test.virtualtabletop.io/PR-1506/pr-test (or any other room on that server)

After merging, a backup will be available at `https://beta.virtualtabletop.io/editor/PR1506-pr-test`.

RaphaelAlvez commented:
@rogerl50 would you mind mixing this with #1237?

I stopped working on that PR because I had trouble with space. Also we wanted to add a color picker for game pieces and a drop-down for spinners.

if not, here's my review on the current setup:
I prefer the current layout in general. I just have some minor suggestions

![image](https://user-images.githubusercontent.com/73476656/210855644-ce17abf9-90dd-4e4e-b90c-eabed098f2f0.png)
this would look better side by side

![image](https://user-images.githubusercontent.com/73476656/210855743-d4e79727-1c78-44ce-8bc4-992039b2ab5c.png)
I believe we decided we could remove those in my PR

![image](https://user-images.githubusercontent.com/73476656/210855910-9c6acf58-3677-4e5f-8f5e-c4266b81647f.png)
Right side has more space

rogerl50 commented:
> @rogerl50 would you mind mixing this with #1237?
> 

> if not, here's my review on the current setup: I prefer the current layout in general. I just have some minor suggestions
When you say *current layout*, do you mean you'd rather not make any of these changes?

rogerl50 commented:
I didn't see #1237. I see no reason to use this PR instead of yours, unless you prefer this layout.

RaphaelAlvez commented:
> I didn't see #1237. I see no reason to use this PR instead of yours, unless you prefer this layout.

No. I prefer your layout.

I was asking if you could help me adding the needed inputs that kept that PR from progressing.

rogerl50 commented:
> > I didn't see #1237. I see no reason to use this PR instead of yours, unless you prefer this layout.
> 
> No. I prefer your layout.
> 
> I was asking if you could help me adding the needed inputs that kept that PR from progressing.

I thought briefly about a color picker and spinner picker, just long enough to realize I had no idea how to do it :)
We probably need more than just a `formField`; at the least we need some way to cancel the operation. Perhaps we can reuse the INPUT code (as @ArnoldSmith86 just did for the scoreboard editor).

ArnoldSmith86 commented:
Keep in mind that I'll start with the new editor soon and that will look completely different anyway. Not saying, I don't want this. Just a heads-up in case you plan to spend a lot of time on this. :smile:

RaphaelAlvez commented:
> Keep in mind that I'll start with the new editor soon and that will look completely different anyway. Not saying, I don't want this. Just a heads-up in case you plan to spend a lot of time on this. 😄

do you realize that this is going to take months again? the inputs would be a great sort term imporovements

rogerl50 commented:
Latest commit adds a popup for spinner values selection. Someone should look at the code. Also, it'd be nice to have the popup show up where the spinner is, but I'm not sure that's possible.

rogerl50 commented:
> Keep in mind that I'll start with the new editor soon and that will look completely different anyway. Not saying, I don't want this. Just a heads-up in case you plan to spend a lot of time on this. 😄

So all of my knowledge of jsonedit.js will be useless? 😢

RaphaelAlvez reviewed (dismissed):
from a quick code review I can't pin point what is the problem but the overlay is being displayed every time I open the room

RaphaelAlvez commented:
> > Keep in mind that I'll start with the new editor soon and that will look completely different anyway. Not saying, I don't want this. Just a heads-up in case you plan to spend a lot of time on this. 😄
> 
> So all of my knowledge of jsonedit.js will be useless? 😢

should be a parallel feature

ArnoldSmith86 commented:
Yeah, I won't touch that a lot. I'll just have to look at top-level commands. Those should be available in the normal editor. But the JSON editor will just be part of that.

rogerl50 commented:
> from a quick code review I can't pin point what is the problem but the overlay is being displayed every time I open the room

Yeah. Displayed and then hidden again. I'll have to take a look. Right now I'm implementing color choosers for the pieces, then I'll clean up the code and commit again, then (by this time it will be tomorrow) I'll try to track down that problem.

RaphaelAlvez commented:
> > from a quick code review I can't pin point what is the problem but the overlay is being displayed every time I open the room
> 
> Yeah. Displayed and then hidden again. I'll have to take a look. Right now I'm implementing color choosers for the pieces, then I'll clean up the code and commit again, then (by this time it will be tomorrow) I'll try to track down that problem.

It would be better if it were outside of the widget click. Somewhere in the top right of the game pieces section.

But I'm happy with whatever you can come up with

rogerl50 commented:
Latest commit pops a color dialog for pieces as well. Also fixes the problem of the edit overlay flashing on load.

There's lots of space now; I haven't bothered to lay things out since there will be new pieces from #1237. @RaphaelAlvez we can merge these in either direction, whichever is easier.

rogerl50 commented:
This will cause TestCafe tests to fail, since those buttons are no longer there...

RaphaelAlvez commented:
> This will cause TestCafe tests to fail, since those buttons are no longer there...

I prefer to close #1237 and bring to changes to here. That PR is pretty old and it will probably be pure code conflicts

rogerl50 commented:
> > This will cause TestCafe tests to fail, since those buttons are no longer there...
> 
> I prefer to close #1237 and bring to changes to here. That PR is pretty old and it will probably be pure code conflicts

 OK, why don't you port that over and then we can deal with the layout. There may be other opportunities for using the color picker once you do that, as well.

RaphaelAlvez commented:
I just saw you changed the structure of you grid. In no way mine was better. I don't even remember why I did stuff back there.

Also I'm not home as soon as I can I'll ring the changes to here. Also I can't see your changes at the moment because my phone doesn't open edit mode.

Definitely happy to close that PR. It's my oldest PR and it always bother me to see it there untouched

rogerl50 commented:
> I just saw you changed the structure of you grid. In no way mine was better. I don't even remember why I did stuff back there.
> 
> Also I'm not home as soon as I can I'll ring the changes to here. Also I can't see your changes at the moment because my phone doesn't open edit mode.
> 
> Definitely happy to close that PR. It's my oldest PR and it always bother me to see it there untouched

I changed it because yours worked and mine really didn't. Maybe we can figure out a cleaner way.

96LawDawg commented:
Not much in the way of substance to add so far, but I like where this is going.  I would suggest adding a header to the color picker like "Choose game piece color".  I think purple is an odd choice for the default color.  I would go with blue or maybe VTT blue as it is more "normal" and more in the center of the color picker (at least on Chrome and FF).  I would also change the default spinner to 6 and set the default number value to start at 6.  But those are just suggestions.  None of that really matters.

rogerl50 commented:
> Not much in the way of substance to add so far, but I like where this is going. I would suggest adding a header to the color picker like "Choose game piece color". I think purple is an odd choice for the default color. I would go with blue or maybe VTT blue as it is more "normal" and more in the center of the color picker (at least on Chrome and FF). I would also change the default spinner to 6 and set the default number value to start at 6. But those are just suggestions. None of that really matters.

Both good ideas. Done.

rogerl50 commented:
If someone knows how to align the four buttons in the Interactive section properly, I'd be grateful. @flackr @robartsd

96LawDawg commented:
> If someone knows how to align the four buttons in the Interactive section properly, I'd be grateful. @flackr @robartsd

They are aligned "properly" in the sense that each group of 2 is centered in the div.  But the buttons are different sizes, so it doesn't look like you want.  Is this what you want?

![Interactive](https://user-images.githubusercontent.com/76912527/211079235-162b4909-96cb-449e-8d06-dc8ab24634ec.png)

You can do that by changing lines 179-180 in room.html to:

```
<button class="ui-button" id="addSeat" style="margin: 0px 17px">Add Seat</button><br>
<button class="ui-button" id="addSeatCounter" style="margin: 0px 17px">+ Counter</button>
```

rogerl50 commented:
I want each button in the bottom row to be centered under the one above it. I was hoping for a solution that didn't involve pixel counting.

ArnoldSmith86 commented:
Put them into a div and use `text-align: center`, maybe?

flackr commented on client/css/editmode.css:
```diff
@@ -1,30 +1,67 @@
 }
+
+#gridContainerAddOverlay > .category > div > button {
+  margin-bottom: 8px;
```
Seems like every .ui-button on this page wants margin-bottom: 8px, maybe just add single selector to do this, e.g.:
```css
#gridContainerAddOverlay .ui-button {
  margin-bottom: 8px;
}
```

Though this makes me wonder why we don't have `margin: 8px;` (i.e. on all sides) of all `.ui-button`.

rogerl50 replied:
Done.

flackr commented on client/css/layout.css:
```diff
@@ -936,6 +936,8 @@ button:disabled:hover {
   cursor: pointer;
   box-shadow: rgba(0, 0, 0, 0.12) 0px 1px 3px 0px, rgba(0, 0, 0, 0.24) 0px 1px 2px 0px;
+  margin-left: 8px;
+  margin-right: 8px;
```
Why not `margin: 8px;` since it seems everywhere this is used you want to separate them by 8px?

rogerl50 replied:
Done.

flackr commented:
> If someone knows how to align the four buttons in the Interactive section properly, I'd be grateful. @flackr @robartsd

You currently have them in two different (i.e. independent) flex rows so they can't align. If you want them aligned you should put them in a grid, table, or even a pair of flex columns (since their heights are the same).

rogerl50 commented on client/css/editmode.css:
```diff
@@ -1,30 +1,67 @@
 }
 
-#addOverlay > .category {
+#gridContainerAddOverlay {
```
This didn't quite work. `box-sizing` seemed to have no effect, but I did need to put `gap: 20px` into this style. In any case, I did remove the wrapper element.

rogerl50 commented on client/css/editmode.css:
```diff
@@ -1,30 +1,67 @@
 
-#addOverlay > .category {
+#gridContainerAddOverlay {
+  display: grid !important;
```
Removing !important resulted in the subsidiary divs being laid out sequentially top to bottom. Did not change this.

rogerl50 commented on client/css/editmode.css:
```diff
@@ -1,30 +1,67 @@
+#gridContainerAddOverlay > .category {
   background: var(--roomColor);
   width: 100%;
+  height: 100%;
```
Done

rogerl50 commented:
@RaphaelAlvez I added stacks of one chip of each value for 2D and 3D. This all seems to work, except that the pile code you put in `cardDefaults` does not kick in for the add widget screen, or when you first put a widget in the room. That code is pretty complicated and I can't figure it out.

ArnoldSmith86 commented:
For the color picker maybe add "inherit from seat X" somehow?

flackr commented on client/js/main.js:
```diff
@@ -88,15 +88,19 @@ export function showOverlay(id, forced) {
     }
+    //Hack to set grid display for add Overlay
+    if (id == 'addOverlay' && style.display !== 'none') {
+      style.display = 'grid'
```
Instead of doing this. In JS you can set style.display to either `'none'` or `''` and then in the css file the display listed there will be used if JS sets `''`.

Alternately we could add a class or attribute for whether the overlay is displayed. I.e. typically what's done is something like this in the css file:
```css
.overlay[hidden] {
  display: none;
}
```

With JS then setting or removing the hidden attribute.

rogerl50 replied:
This requires changes for every overlay if I'm not mistaken. I'm kind of reluctant to take that on in this fairly simple PR.

flackr replied:
All of the overlays are of class `.overlay` so you could have two rules cover all of them:
```css
overlay {
  display: flex;
}

#addOverlay {
  display: grid;
}
```

Alternately if you go with the HTML attribute to toggle visibility,
```css
overlay[hidden] {
  display: none !important;
}

overlay {
  display: flex;
}

#addOverlay {
  display: grid;
}
```

rogerl50 replied:
While this is a really clean design, I'm just not comfortable putting it into this PR. Beyond the CSS changes, which are indeed fairly easy (though they require some other changes in `room.html` and a couple of overlay css files), there is at least one other place in the code where `display: none` is tested for (this is in the handler for a click on the toolbar buttons). If this is to be done, it should be its own PR.

flackr replied:
Okay, instead of this override here, can you move 'flex' or 'grid' into a local var so it's not an override after setting the other style. E.g. change lines 91-92 to:

```js
const activeDisplay = id == 'addOverlay' ? 'grid' : 'flex';
style.display = !forced && style.display !== 'none' ? 'none' : activeDisplay;
overlayActive = style.display !== 'none';
```

rogerl50 replied:
Sold. Still unfortunate, but cleaner.

rogerl50 commented:
This is ready to merge once @flackr dismisses his requested changes (which have been implemented) and we get an approval or two.

96LawDawg commented:
The 3-D poker chip needed a dropTarget, so I fixed that.  Because the chips are a deck, I thought we should treat it like the regular deck and I added a recall and sort button.  If you don't like it, it can be deleted/reverted.

96LawDawg reviewed (approved):
Approved through test.  I've clicked every widget in the overlay.  They all work.

rogerl50 commented:
> The 3-D poker chip needed a dropTarget, so I fixed that. Because the chips are a deck, I thought we should treat it like the regular deck and I added a recall and sort button. If you don't like it, it can be deleted/reverted.

Is that the standard recall and shuffle icon now?

96LawDawg commented:
> Is that the standard recall and shuffle icon now?

Recall and _sort_.  We could just do a recall, but makes sense to me to sort them.  Again, I don't really care.

RaphaelAlvez commented:
> > Is that the standard recall and shuffle icon now?
> 
> Recall and _sort_.  We could just do a recall, but makes sense to me to sort them.  Again, I don't really care.

I would prefer recall **&** sort but I think it's alright

It makes no sense to shuffle this deck and it can be a good example

96LawDawg commented:
> I would prefer recall **&** sort but I think it's alright
> 
> I mean to just add an & but that will require a new hash abs maybe not worth

Thanks for clarifying. But that isn't even possible.  We can't mix text and symbols.  @mousewax never got that working.  It is possible to have text and 1 image, but unless we make a special image for this (we should not), then it is either what we have, or just use text.  Obviously, I'm a fan of using what we have.

flackr reviewed (approved):
Looks good with removal of overriding display hack.
